### PR TITLE
Versions expiration cronjob

### PR DESCRIPTION
**Pull request by VicDeo**
The specification is in https://github.com/owncloud/core/issues/16301
- [x] https://github.com/owncloud/core/pull/18368 should go first

### CODE REVIEW

**Comment by karlitschek** [comment]
nice :+1:

**Comment by VicDeo** [comment]
rebased

File: apps/files_versions/appinfo/app.php [hunk 1]
```diff
@@ -19,6 +19,15 @@
...
+
+$app = new Application();
+$container = $app->getContainer();
+
+\OCP\Util::addStyle('files_versions', 'versions');
 
 \OCA\Files_Versions\Hooks::connectHooks();
+
+// Cron job for deleting expired trash items
+$container->getServer()->getJobList()->add('OCA\Files_Versions\BackgroundJob\ExpireVersions');
```
Comment (nickvergessen) [inline apps/files_versions/appinfo/app.php]: This runs a query on each page run! :zap: 
Please move to install.php and update.php
See #18781

File: apps/files_versions/lib/backgroundjob/expireversions.php [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+		\OC_Util::tearDownFS();
+	}
+
+	/**
+	 * Act on behalf on trash item owner
+	 * @param string $user
+	 * @return boolean
+	 */
+	private function setupFS($user){
+		$userManager = \OC::$server->getUserManager();
```
Comment (nickvergessen) [inline apps/files_versions/lib/backgroundjob/expireversions.php]: Use your `fixDIForJobs()` method?

File: apps/files_versions/lib/backgroundjob/expireversions.php [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+		$this->expiration = $application->getContainer()->query('Expiration');
+	}
+
+	protected function run($argument) {
+		$maxAge = $this->expiration->getMaxAgeAsTimestamp();
+		if (!$maxAge) {
+			return;
+		}
+
+		$userManager = \OC::$server->getUserManager();
```
Comment (nickvergessen) [inline apps/files_versions/lib/backgroundjob/expireversions.php]: Use your `fixDIForJobs()` method?

File: apps/files_versions/tests/expiration.php [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License, version 3,
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>
+ *
+ */
+
+namespace OCA\Files_Versions;
```
Comment (nickvergessen) [inline apps/files_versions/tests/expiration.php]: `OCA\Files_Versions\Tests`

File: apps/files_versions/tests/expiration.php [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+
+class Expiration_Test extends \PHPUnit_Framework_TestCase {
```
Comment (nickvergessen) [inline apps/files_versions/tests/expiration.php]: please extend `\Test\TestCase`

File: apps/files_versions/lib/backgroundjob/expireversions.php [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+
+	/**
+	 * @var Expiration
+	 */
+	private $expiration;
+
+	/**
+	 * @var IDBConnection
+	 */
+	private $dbConnection;
```
Comment (schiesbn) [inline apps/files_versions/lib/backgroundjob/expireversions.php]: seems to be unused

File: apps/files_versions/lib/backgroundjob/expireversions.php [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+		if (!$this->userManager->userExists($user)) {
+			return false;
+		}
+
+		\OC_Util::tearDownFS();
+		\OC_Util::setupFS($user);
+
+		return true;
+	}
+}
```
Comment (schiesbn) [inline apps/files_versions/lib/backgroundjob/expireversions.php]: missing new line

**Comment by schiesbn** [comment]
code looks good beside the small things I mentioned above :+1:

**Comment by PVince81** [comment]
Do we still need the old "expire versions" one-shot background task `OCA\Files_Versions\Command\Expire`. Remove it ?

File: apps/files_versions/appinfo/app.php [hunk 8]
```diff
@@ -19,6 +19,12 @@
  * along with this program.  If not, see <http://www.gnu.org/licenses/>
  *
  */
-OCP\Util::addStyle('files_versions', 'versions');
+
+namespace OCA\Files_Versions\AppInfo;
+
+$app = new Application();
+$container = $app->getContainer();
```
Comment (nickvergessen) [inline apps/files_versions/appinfo/app.php]: No need to init the container? you don't use it?

**Comment by VicDeo** [comment]
Rebased. Squashed.

**Comment by VicDeo** [comment]
@nickvergessen @schiesbn comments are addressed.

**Comment by PVince81** [comment]
- [ ] TODO after merge: raise separate ticket for discussing "remove expiration command" as it might be obsoleted by this cron job

Now just wait for CI...

File: apps/files_versions/lib/backgroundjob/expireversions.php [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+		$isFSready = false;
+		foreach ($users as $user) {
+			$uid = $user->getUID();
+			if (!$isFSready) {
+				if (!$this->setupFS($uid)) {
+					continue;
+				}
+				$isFSready = true;
+			}
+			Storage::expireOlderThan($uid);
```
Comment (nickvergessen) [inline apps/files_versions/lib/backgroundjob/expireversions.php]: Expire older then username?! The method name either does not do what I think, or the parameter is a wrong one?

File: apps/files_versions/lib/storage.php [hunk 10]
```diff
@@ -403,6 +403,34 @@ public static function getVersions($uid, $filename, $userFullPath = '') {
 		return $versions;
 	}
 
+	public static function expireOlderThan($uid){
```
Comment (nickvergessen) [inline apps/files_versions/lib/storage.php]: See comment above

Comment (nickvergessen) [inline apps/files_versions/lib/storage.php]: Missing doc block

File: apps/files_versions/lib/backgroundjob/expireversions.php (hunk 9 above)
Comment (nickvergessen) [inline apps/files_versions/lib/backgroundjob/expireversions.php]: Maybe more something like `expireOlderVersionsForUser()` ?

**Comment by VicDeo** [comment]
@PVince81
 this cronjob expires only items that are older than maximal retention time.
It doesn't do version thinning like `OCA\Files_Versions\Command\Expire` does.
The former used when there is non-auto max retention, the latter is for `auto` mode.

**Comment by VicDeo** [comment]
@nickvergessen updated

**Comment by PVince81** [comment]
@VicDeo thanks for the clarification, makes sense

**Pull request merged**